[@kbn/eslint-plugin-eslint] Add ESLint rule that requires `apiClient` usage in Scout `apiTest`

steliosmavro opened the pull request:
## Summary

Adds a new ESLint rule `scout_require_api_client_in_api_test` that enforces the use of the `apiClient` fixture in `apiTest` functions to ensure API endpoints are properly tested.

## What it does

When writing API tests with `apiTest`, developers must use the `apiClient` fixture to make HTTP calls. The rule reports an error if `apiClient` is destructured but never used.
**Invalid:**
```
apiTest('test', async ({ apiClient }) => {
  console.log('no api call'); // ❌ apiClient not used
});
```
**Valid:**
```
apiTest('test', async ({ apiClient }) => {
  await apiClient.get('/foo'); // ✅ apiClient used
});
```

## Features

- Detects direct usage, aliases, and variable assignments
- Tracks usage through helper functions and closures
- Supports `eslint-disable-next-line` overrides

## Changes

- Added rule implementation and comprehensive test suite (21 test cases)
- Registered rule in plugin and enabled as error for Scout test files

<!--ONMERGE {"backportTargets":["8.19","9.1","9.2"]} ONMERGE-->

steliosmavro committed:
[kbn-eslint-plugin-eslint] add scout_api_test_encourage_api_client rule

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/368714)
* Commit: 6f780c71da90abd921848454e5078b961a7d6c8e

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
<details><summary>Unknown metric groups</summary>

#### ESLint disabled in files
| id | [before](https://github.com/elastic/kibana/commit/e161e4e37b15d0987ea55c7c7d1c514db6b739da) | [after](https://github.com/elastic/kibana/commit/6f780c71da90abd921848454e5078b961a7d6c8e) | diff |
| --- | --- | --- | --- |
| `@kbn/eslint-plugin-eslint` | 0 | 2 | +2 |

#### ESLint disabled line counts
| id | [before](https://github.com/elastic/kibana/commit/e161e4e37b15d0987ea55c7c7d1c514db6b739da) | [after](https://github.com/elastic/kibana/commit/6f780c71da90abd921848454e5078b961a7d6c8e) | diff |
| --- | --- | --- | --- |
| `@kbn/eslint-plugin-eslint` | 0 | 2 | +2 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/e161e4e37b15d0987ea55c7c7d1c514db6b739da) | [after](https://github.com/elastic/kibana/commit/6f780c71da90abd921848454e5078b961a7d6c8e) | diff |
| --- | --- | --- | --- |
| `@kbn/eslint-plugin-eslint` | 0 | 4 | +4 |


</details>


<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368714","number":368714,"commit":"6f780c71da90abd921848454e5078b961a7d6c8e"}],"number":368714}
buildkite-pr-comment-->

steliosmavro committed:
Improvements:
- Add test case
- Update rule according to new test
- Rename rule and files
- Throw error instead of warning

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/369591)
* Commit: 4ecedc8b1ed3d4ac0d39ac6806b1d6333cdfb7e1

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
<details><summary>Unknown metric groups</summary>

#### ESLint disabled in files
| id | [before](https://github.com/elastic/kibana/commit/e161e4e37b15d0987ea55c7c7d1c514db6b739da) | [after](https://github.com/elastic/kibana/commit/4ecedc8b1ed3d4ac0d39ac6806b1d6333cdfb7e1) | diff |
| --- | --- | --- | --- |
| `@kbn/eslint-plugin-eslint` | 0 | 2 | +2 |

#### ESLint disabled line counts
| id | [before](https://github.com/elastic/kibana/commit/e161e4e37b15d0987ea55c7c7d1c514db6b739da) | [after](https://github.com/elastic/kibana/commit/4ecedc8b1ed3d4ac0d39ac6806b1d6333cdfb7e1) | diff |
| --- | --- | --- | --- |
| `@kbn/eslint-plugin-eslint` | 0 | 2 | +2 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/e161e4e37b15d0987ea55c7c7d1c514db6b739da) | [after](https://github.com/elastic/kibana/commit/4ecedc8b1ed3d4ac0d39ac6806b1d6333cdfb7e1) | diff |
| --- | --- | --- | --- |
| `@kbn/eslint-plugin-eslint` | 0 | 4 | +4 |


</details>


### History
* :green_heart: [Build #368714](https://buildkite.com/elastic/kibana-pull-request/builds/368714) succeeded 6f780c71da90abd921848454e5078b961a7d6c8e

cc @steliosmavro

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/369591","number":369591,"commit":"4ecedc8b1ed3d4ac0d39ac6806b1d6333cdfb7e1"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368714","number":368714,"commit":"6f780c71da90abd921848454e5078b961a7d6c8e"}],"number":369591}
buildkite-pr-comment-->

csr reviewed (approved):
LGTM. Nice job! 👏  We might want to `backport:all-open` to introduce this rule in all active version branches. Check with DIma/team.

dmlemeshko reviewed (approved):
Great job, I went through tests and it seems like your rule covers exactly what we need 🚀

steliosmavro committed:
Merge branch 'main' into eslint-apiClient-rule

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.19|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/245708">](https://github.com/elastic/kibana/pull/245708)|
|✅|9.1|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/245709">](https://github.com/elastic/kibana/pull/245709)|
|✅|9.2|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/245710">](https://github.com/elastic/kibana/pull/245710)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->
